app/uk/gov/hmrc/alcoholdutyreturns/service/SchemaValidationService.scala

```diff
@@ -0,0 +1,145 @@
+  }
+
+  import SchemaType._
+
+  private val schemaFileNames: Map[SchemaType, String] = Map(SubmitReturn -> "submitReturnSchema.json")
+
+  private val jsonSchemaFactory = JsonSchemaFactory
+    .getInstance(
+      SpecVersion.VersionFlag.V202012,
+      enrichAsJavaConsumer((builder: JsonSchemaFactory.Builder) =>
+        builder.metaSchema(OpenApi30.getInstance()).defaultMetaSchemaIri(OpenApi30.getInstance().getIri)
+      ).asJava
+    )
+
+  private def loadSchema(schemaFileName: String): Option[JsonSchema] = {
+    val is = env.classLoader.getResourceAsStream(s"${appConfig.schemaDir}/$schemaFileName")
+
+    if (is != null) {
+      val maybeSchema = Try {
+        jsonSchemaFactory.getSchema(is)
+      } match {
+        case Success(schema) =>
+          logger.info(s"Successfully loaded json schema $schemaFileName")
+          Some(schema)
+        case Failure(_)      =>
+          logger.error(s"Unable to load json schema $schemaFileName")
+          None
+      }
+
+      is.close()
+      maybeSchema
+    } else {
+      None
+    }
+  }
+
+  private val schemas: Map[SchemaType, JsonSchema] = {
+    val loadAttemptedSchemas = SchemaType.values.map { schema =>
+      val fileName = schemaFileNames.getOrElse(
+        schema, {
+          logger.error(s"Cannot find filename for schema ${schema.entryName}")
+          throw new RuntimeException(s"Cannot find filename for schema ${schema.entryName}")
+        }
+      )
+      loadSchema(fileName).map(schema -> _)
+    }
+    val numSchemas           = loadAttemptedSchemas.length
+    val schemasLoaded        = loadAttemptedSchemas.flatten
+    val numSchemasLoaded     = schemasLoaded.length
+
+    logger.info(s"Validation schemas $numSchemas of $numSchemasLoaded loaded")
+
+    if (numSchemas != numSchemasLoaded) {
+      logger.error(s"Only $numSchemas of $numSchemasLoaded schemas loaded - unable to start server")
+      throw new RuntimeException("Schema loading failed")
+    } else {
+      logger.info(s"All validation schemas($numSchemas) were loaded")
+      schemasLoaded.toMap
+    }
+  }
+
+  private def selectSchema[T](obj: T): Either[Throwable, JsonSchema] = obj match {
+    case _ if obj.isInstanceOf[ReturnCreate] =>
+      schemas
+        .get(SubmitReturn)
+        .fold[Either[Throwable, JsonSchema]](
+          Left(new RuntimeException(s"Schema ${SubmitReturn.entryName} should have been loaded"))
+        )(schema => Right(schema))
+    case _                                   =>
+      logger.error(s"Attempting schema validation against unsupported object type ${obj.getClass.getTypeName}")
+      Left(
+        new RuntimeException(
+          s"Attempting schema validation against unsupported object type ${obj.getClass.getTypeName}"
+        )
+      )
+  }
+
+  /**
+    * @return true if should continue, false if to fail with e.g. BadRequest
+    */
+  def validateAgainstSchema[T](obj: T)(implicit writes: Writes[T]): Boolean =
```

Comment:
This should just receive a schema path and a JsValue and validate it. It shouldn't worry about converting an object to a class using implicit writes. It should be done before coming to this method. 

It is also unnecessarily complex. We don't need the selectSchema there, like I mentioned above, it could be just a path.

Antyways this is working and considering timelines, we'll have to refactor this later

Reply:
It was to decouple from the implementation. The caller doesn't need to know the path of the schema, nor does it need to have to know it needs a Jackson parser to load it rather than a play one (can one convert from JsValue to the correct type?)

Reply:
Also the schemas have to be loaded here anyway, otherwise we can't fail at start-up.